aaronthangnguyen/cen4010-swe1
# Add `get books by rating` Endpoint

### Summary

We've elevated our book management system by introducing one pivotal `GET` endpoints. This endpoint will help the users to be able to retrieve book lists from a particular rate and higher. This enhancement optimizes the browsing experience by providing tailored book recommendations and highlighting trending literature, ensuring our readers always have access to the best of our catalog.

### Changes

#### Change in Book.java:
+ Add double averageRating include getter, setter
#### Change in BookController.java:
+ Add exception.PublisherNotFoundException
+ Add GetMapping("/by-rating")
+ Rename DiscountDTO to discount
#### Change in BookService.java:
+ import stream.Collectors
+ import RateAndCommentRepository
+ add private RateAndCommentRepository
+ add this.rateAndCommentRepository = rateAndCommentRepository;
+ add public List<Book> getBooksByRatingOrHigher
#### Change in Discount.java:
+ rename DiscountDTO to Discount
#### Change in Makefile:
+ Update it can run on Windows Terminal
#### Change in PublisherNotFoundException.java:
+ Moved to exception folder
#### Change in RateAndComment.java:
+ Modify type of bookID from Long to UUID
#### Change in RateAndCommentController.java:
+ import UUID
+ Long bookID to UUID bookID
#### Change in RateAndCommentRepository.java:
+ Long bookID to UUID bookID
+ Add new Repo for findBookIdsByRatingOrHigher
#### Change in RateAndCommentService.java:
+ import UUID
+ Long bookID to UUID bookID

### Endpoint Details
- Purpose: Allows users or administrators to retrieve a list of books that has a particular rating or higher.
- Endpoint: `api/books/by-rating?rating=n` as `n` is a number that has the type of double
- HTTP Method: `GET`
- Response: Returns a JSON array containing detailed information about all books with the same rating or higher in the system. Each book entry includes essential details such as the book's ISBN, title, description, price, author, genre, publisher, publication year, the number of copies sold, and rating.
    Example Response:
```json
[
 {
  "id":"15a65083-89c0-454e-af81-08c2b2d2a6f4",
  "isbn":"978-7890123456",
  "name":"The Scientific Revolution",
  "description":"Explore the scientific discoveries that shaped our world.",
  "price":-179.91,"authorId":"223e4567-e89b-12d3-a456-556642440000",
  "genre":"Non-Fiction",
  "publisher":"Discovery Publications",
  "yearPublished":2023,
  "copiesSold":700,
  "averageRating":4.5
 }
]
```

Test Cases
- Retrieve the list of all books by rating.
- Check for an empty list if there are no books in the system.
- Verify the response structure and data accuracy.

How to Test
- Pull the branch to your local setup.
- Run the application.
- Use the provided curl command or any API tool like Postman to test the endpoint.

Example
```cmd
curl http://localhost:8080/api/books/by-rating?rating=4.0
```
